### ISSUE TITLE
Polonius fails to infer lifetimes of borrows

### ISSUE THREAD

**Issue by zeenix** [rust-lang/rust#134554]
I tried this code:

```rust
struct Foo([u8; 10]);

impl Foo {
    fn read_non_empty<'c, P>(&'c mut self) -> P
    where
        P: serde::Deserialize<'c> + std::string::ToString,
    {
        loop {
            let b: P = self.read();
            if !b.to_string().is_empty() {
                return b;
            }
        }
    }

    fn read<'c, P>(&'c mut self) -> P
    where
        P: serde::Deserialize<'c>,
    {
        unimplemented!();
    }
}
```

I expected to see this happen: Compilation to succeed.

Instead, this happened: Got an error:

```rust
❯  RUSTFLAGS='-Zpolonius' cargo +nightly c --tests
    Checking tmp-scdnga v0.1.0 (/home/zeenix/.cache/cargo-temp/tmp-sCDnga)

error[E0499]: cannot borrow `*self` as mutable more than once at a time
 --> src/main.rs:9:24
  |
4 |     fn read_non_empty<'c, P>(&'c mut self) -> P
  |                       -- lifetime `'c` defined here
...
9 |             let b: P = self.read();
  |                        ^^^^-------
  |                        |
  |                        `*self` was mutably borrowed here in the previous iteration of the loop
  |                        argument requires that `*self` is borrowed for `'c`

For more information about this error, try `rustc --explain E0499`.
error: could not compile `tmp-scdnga` (bin "tmp-scdnga" test) due to 1 previous error
```

If I modify the code to not use a loop, I still get the same:

```rust
    fn read_non_empty<'c, P>(&'c mut self) -> P
    where
        P: serde::Deserialize<'c> + std::string::ToString,
    {
        {
            let b: P = self.read();
            if !b.to_string().is_empty() {
                return b;
            }
        }

        {
            let b: P = self.read();
            if !b.to_string().is_empty() {
                return b;
            }
        }

        unimplemented!();
    }
```

### Meta

`rustc --version --verbose`:
```
rustc 1.85.0-nightly (9e136a30a 2024-12-19)
binary: rustc
commit-hash: 9e136a30a965bf4e63f03095c57df7257bf96fd6
commit-date: 2024-12-19
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

**Comment by bjorn3** [comment]
`-Zpolonius` enables the legacy implementation of polonius. `-Zpolonius=next` enables the new from scratch implementation. Polonius next is not yet complete, but legacy polonius has several known issues that polonius next is supposed to fix AFAIK.

**Comment by zeenix** [comment]
> Polonius next is not yet complete, but legacy polonius has several known issues that polonius next is supposed to fix AFAIK.

Gotcha. Just for the record, I get an error with `RUSTFLAGS='-Zpolonius=next'` as well, which I guess is expected since polonius-next is incomplete.

**Comment by zeenix** [comment]
> Your `read` call is an exclusive borrow for `'c`, that lasts for the rest of the function since you return the value

But why should me returning the value, make the borrow last for the entire scope of the function? 🤔 If I return the value, nothing else in the function can access it anymore. If I do not return the value, the scope of the borrow should ensure that it's dropped (explicitly dropping doesn't help here either).

> fn read_non_empty(foo: &mut u32) {
>     let a = read(foo);
>     let _ = read(foo);
>     drop(a);
> }

but that's not my use case. I need to return the value if it satisfied a certain condition, otherwise keep reading it.

**Comment by zeenix** [comment]
> > fn read_non_empty(foo: &mut u32) {
> > let a = read(foo);
> > let _ = read(foo);
> > drop(a);
> > }
> 
> but that's not my use case. I need to return the value if it satisfied a certain condition, otherwise keep reading it.

Ah, you meant to reduce the reproducer. Sorry, I misunderstood.

**Comment by KiChjang** [comment]
> But why should me returning the value, make the borrow last for the entire scope of the function? 🤔 If I return the value, nothing else in the function can access it anymore. If I do not return the value, the scope of the borrow should ensure that it's dropped (explicitly dropping doesn't help here either).

Because your where clause demands it: `read_non_empty<'c, P>(&'c mut self) -> P where P: Deserialize<'c>`. In other words, the where clause contains a contract that `P` may contain a borrow of `self` that lasts for `'c`. Now look at what `self.read()` does -- it also has its own lifetime parameter (call it `'x`) and where clause `RetTy: Deserialize<'x>`. Since `b` ends up being `P` (via both your type annotation and `return b`), the call to `self.read()` is resolved with `RetTy = P` and `'x = 'c`, per your where clause. 

This means the moment you call `self.read()`, the call makes `self` mutably borrowed for the rest of the function. We can see this happening in a smaller example:

```rust
fn f<'c>(x: &'c mut i32) {
    let r: &'c mut i32 = &mut *x;
    drop(r);
    *x = 5; // error: cannot use `*x` because it was mutably borrowed
}
```

`r` is explicitly dropped, provably gone, and yet `x` still cannot be used, because dropping `r` doesn't kill the existing loan on `x` as it's intended to last as long as `'c`. You may ask where this requirement comes from, and it's from writing `let r: &'c mut i32 = &mut *x;`.

So now you may have a eureka moment -- I can simply just remove the `'c` requirement, and the loan on `x` will be gone! That's exactly right: if you try and remove the `'c` from the type ascription of `r`, the error then vanishes, because now the borrow made by `r` is inferred to be an anonymous one that doesn't span `'c`. However, this doesn't solve your case -- you actually want `P` to satisfy `Deserialize<'c>`, because you can then return `b` directly after the check without any intermediate clones or copies for the check.

To really build the intuition of understanding why merely calling `self.read()` can hold the self reference alive for `'c`, observe that the `'de` lifetime in `Deserialize<'de>` [eventually gets used by `visit_borrowed_str` in the deserializer](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L1543). Since `Deserialize` is a trait whose code you don't control, there's no guarantee by the compiler that you won't feed a custom `Deserializer` to the [`deserialize` trait function](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L561) and make a custom `visit_borrowed_str` impl that copies the `&'de str` reference and stores it somewhere else. Hence, the compiler assumes the worst case where `self` is going to be held alive for `'c` by `self.read()` due to the trait bound `P: Deserialize<'c>`, because inside of the trait hierarchy introduced by  `Deserialize<'c>`, a `&'c str` reference is being handed out in a trait function to custom code.

Hence @lqd's reduction actually diverges from the original -- his version elides the lifetime, but a faithful reduction requires the explicit return of `b` (or an annotation of `b: &'c u32`):

```rust
fn read_non_empty<'c>(foo: &'c mut u32) -> &'c u32 {
    let b = read(foo);
    let _ = read(foo);
    b
}

fn read<'a>(_: &'a mut u32) -> &'a u32 {
    unimplemented!()
}
```

Link: https://godbolt.org/z/KMfG8b3jj

This version errors under NLL, `-Zpolonius`, and `-Zpolonius=next` alike, with no returns anywhere. In @lqd's elided version the lifetime is fresh per call, so overwriting `b` each iteration lands in #47680 territory and polonius accepts it, but the generic original has no such freedom: the where clause pins every call to `'c`, so returning, dropping, or overwriting `b` cannot help even in principle.

So in conclusion, the original code is probably fine in most use cases, since `Deserialize` is most likely going to be derived, but it cannot be generalized to _all_ cases, since the `P: Deserialize<'c>` requirement is where the unsoundness may arise in some cases. The borrow checker is correct to reject the code as-is.

**Comment by KiChjang** [comment]
The code snippet provided by @lengyijun compiles, but it's worth understanding what it does -- `P: for<'a> Deserialize<'a>` is a universal quantification for _any_ choice of `'a`, and it's actually quite a problem because now `'a` doesn't have any relationship at all with `'c` when what we really want is `'c: 'a`. In practice, this means `P` can only be `DeserializeOwned` and not a borrow of `self`, since `DeserializedOwned` has a blanket implementation for `T: for<'de> Deserialize<'de>`.

**Comment by zeenix** [comment]
> Because your where clause demands it: `read_non_empty<'c, P>(&'c mut self) -> P where P: Deserialize<'c>`. In other words, the where clause contains a contract that `P` may contain a borrow of `self` that lasts for `'c`.

Thanks for explaining but I know that part and that `read` is borrowing `'c` but the return value of `read` is out of scope so anything that is borrowed by `read` and/or its return value should no longer be considered borrowed after the scope ends:

```rust
        {
            let b: P = self.read();
            if !b.to_string().is_empty() {
                return b;
            }
        } // After this point neither `self`, nor `b` should be borrowed anymore.
```

**Comment by KiChjang** [comment]
@zeenix `read` itself calls functions on `Deserialize<'c>`, which hands out references that may last for `'c`, which is what's keeping `self` alive. Even if `b` is out of scope, that function in `Deserialize<'c>` may have stored that reference somewhere else.

**Comment by zeenix** [comment]
> [@zeenix](https://github.com/zeenix) `read` itself calls functions on `Deserialize<'c>`, which hands out references that may last for `'c`, which is what's keeping `self` alive. Even if `b` is out of scope, that function in `Deserialize<'c>` may have stored that reference somewhere else.

Like where and how? All the output of `read` is gone out of the scope and hence all the borrows are gone too.

**Comment by KiChjang** [comment]
> > [@zeenix](https://github.com/zeenix) `read` itself calls functions on `Deserialize<'c>`, which hands out references that may last for `'c`, which is what's keeping `self` alive. Even if `b` is out of scope, that function in `Deserialize<'c>` may have stored that reference somewhere else.
> 
> Like where and how? All the output of `read` is gone out of the scope and hence all the borrows are gone too.

They aren't necessarily gone. Consider where exactly the `'c` gets used in `Deserialize<'c>` -- it first gets threaded through to [`Deserializer<'c>` via the `deserialize` function](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L563), goes through [another layer via `Visitor<'c>`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L961), and lands at either [`visit_borrowed_str`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L1543) or [`visit_borrowed_bytes`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L1602). It's in these two functions where you see the arguments handing out either `&'c str` or `&'c [u8]` to implementors.

Now think about how `read` even gets to create that `P`: somewhere in its body, it _must_ have used a deserializer to construct it, thereby invoking any of the `Visitor` trait functions to do so. Since `Visitor` is a trait whose implementation that you don't control, any code that implements it may choose to store the `&'c str` or `&'c [u8]` reference somewhere, which ultimately is what keeps the `&'c mut self` alive, in addition to the return values that may also store references to `self`. Hence, it's the moment that you call `self.read()` where the `self` reference is kept alive, even if you let the return value of `self.read()` go out of scope.

**Comment by zeenix** [comment]
> They aren't necessarily gone. Consider where exactly the `'c` gets used in `Deserialize<'c>` -- it first gets threaded through to [`Deserializer<'c>` via the `deserialize` function](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L563), goes through [another layer via `Visitor<'c>`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L961), and lands at either [`visit_borrowed_str`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L1543) or [`visit_borrowed_bytes`](https://github.com/serde-rs/serde/blob/1023d077510b4aef36a41ef56fdb7798568a2654/serde_core/src/de/mod.rs#L1602). It's in these two functions where you see the arguments handing out either `&'c str` or `&'c [u8]` to implementors.

Thanks but you're still describing `read` and how the borrowing happens there. The `read` function definitely borrows from `'c` and returns that through its `Deserialize<'c>` return type. That part is super clear. What is not clear is how dropping of that return type, does not also drop the borrow along with it. This question is purely about the using code. How the borrow occurs in the first place inside `read` and returns is not relevant at all AFAICT.

**Comment by KiChjang** [comment]
@zeenix I have already answered you very concretely 3 times and you have missed the part that matters the most: 

> Now think about how `read` even gets to create that `P`: somewhere in its body, it _must_ have used a deserializer to construct it, thereby invoking any of the `Visitor` trait functions to do so. Since `Visitor` is a trait whose implementation that you don't control, any code that implements it may choose to store the `&'c str` or `&'c [u8]` reference somewhere, which ultimately is what keeps the `&'c mut self` alive, in addition to the return values that may also store references to `self`. Hence, it's the moment that you call `self.read()` where the `self` reference is kept alive, even if you let the return value of `self.read()` go out of scope.

You do _not_ get to control what happens inside of `visit_borrowed_str`, so the compiler _must_ assume the worst case and that the reference is stored somewhere else you don't control, i.e. you don't own that part of the data and hence you can't simply just drop it, nor is it returned as a value anywhere in `read`, e.g. an implementor of `Visitor` can simply store the reference in its own state, keeping `self` alive.

I think you're assuming that only the returned `P` value can call `deserialize`, but that isn't true -- during the construction of `P`, a deserializer _must_ have been invoked to give you a `P`. Otherwise, if `read` doesn't deserialize `P` in any way, then it has no way to construct `P` as it's an opaque type.

**Comment by zeenix** [comment]
> Since `Visitor` is a trait whose implementation that you don't control, any code that implements it may choose to store the `&'c str` or `&'c [u8]` reference somewhere

But where and how (exactly what I asked before)? The location it's stored in has to be part of the return value and if it's not then that can be assumed to be a transient borrow (and hence irrelevant). That's why `Deserialize` is borrowing `'c` to indicate to the compiler that the return value has the lifetime borrowed somewhere inside it. You can't just store a reference. You'll have to use unsafe code to store the reference as a pointer to be able to escape one of these 2 possibilties and that's not something the compiler needs to worry about.

**Comment by KiChjang** [comment]
Since `'c` can be any lifetime, we can set `'c = 'static`, and everything you thought wasn't possible is now completely so, such as spawning a new thread and passing the static reference to its closure. Note that this can be done without any unsafe code.

**Comment by zeenix** [comment]
> Since `'c` can be any lifetime, we can set `'c = 'static`, and everything you thought wasn't possible is now completely so

but `'c` is not `static` and it can't be assumed to be static.

**Comment by KiChjang** [comment]
> but `'c` is not `static` and it can't be assumed to be static.

Yes it can, that's what generic code means.

If you accept `T` as a generic parameter, then anyone can simply just put in a `String`, a `bool`, or even a `()` as a type that satisfies `T`, as long as `T` doesn't have any trait bounds to it.

In the same manner, when you have `'c` as a lifetime parameter, then anyone can supply a `'static` legally. Your code must work for _all_ choices of `'c`, not just _some_ choices of `'c` that you get to pick and choose -- you can't just ban `'static` from being supplied, so your code _must_ also work for `'static` lifetimes. And even more specifically to `'static`: it's considered the subtype of _all_ possible lifetimes, so whenever _any_ lifetime parameter is expected in a covariant context, it is _always_ legal to supply a `'static`.

**Comment by zeenix** [comment]
> then anyone can supply a 'static legally. 

Of course but that's not happening here.

I believe I already explained how it's known to the compiler that all borrows are dropped after the scope (or the loop iteration) and hence the rest doesn't (or rather shouldn't) matter.

**Comment by KiChjang** [comment]
> Of course but that's not happening here.

Yes it is, you are writing generic code, so it must work for `'static` as well. You can't just willfully ignore it just because it doesn't appear in the use cases you're familiar with.

I believe I have sufficiently explained how a reference can still escape, why it matters, and why the compiler is working as intended. Absent a compelling reason to keep this issue open, I suggest we close it as resolved.

**Comment by zeenix** [comment]
> I believe I have sufficiently explained how a reference can still escape

You have not, sorry. A borrow in a function is either temporary or returned in the output. If you could just safely store it anywhere, all lifetimes on function args would have to be assumed to be `'static`. So at least theoretically, my example code above should work.

@LeoniePhiline (in a matrix channel) explained to me in great detail the reason this can't work on a practical level and I understand why this would be very challenging for the borrowchecker to reason with.

> I suggest we close it as resolved.

If people working on polonius, agree this is not something they'll be able fix, sure thing.

**Comment by KiChjang** [comment]
> If you could just safely store it anywhere, all lifetimes on function args would have to be assumed to be `'static`.

What you said is a different formulation of what I said. You are saying "if I hand you a specific lifetime `'c`, then it cannot suddenly be upgraded to `'static`", and that part we are in agreement. What I am saying is instead "the specific lifetime of `'c` _is_ `'static`", thereby it _can_ escape.

> I understand why this would be very challenging for the borrowchecker to reason with.

You're asking for borrowck to do a full-program analysis, and is very hard to do in practice, not to mention how it would definitely balloon compile times. In practice, borrowck has to work only on the type-level function signatures and type ascriptions to limit the complexity. In your case, a properly documented unsafe block where you explicitly state that the incoming lifetime cannot be static will suffice.
